packages/flutter/lib/src/material/input_decorator.dart

```diff
@@ -1006,21 +1027,27 @@ class _RenderDecoration extends RenderBox {
       ? maxContainerHeight
       : math.min(contentHeight, maxContainerHeight);
 
-    // Always position the prefix/suffix in the same place (baseline).
+    // Try to consider the prefix/suffix as part of the text when aligning it.
+    // If the prefix/suffix overflows however, allow it to extend outside of the
+    // input and align the remaining part of the text and prefix/suffix.
     final double overflow = math.max(0, contentHeight - maxContainerHeight);
-    final double baselineAdjustment = fixAboveInput - overflow;
+    final double textAlignVerticalFactor = (textAlignVertical.y + 1.0) / 2.0;
```

Comment:
It's not obvious (to me) what the `+ 1.0` and the `1 - ` adjustments are for here and on the next line.